### ISSUE TITLE
Merge filter fusion and `CanonicalizeMfp`

### ISSUE THREAD

**Issue by ggevay** [MaterializeInc/materialize#17142]
Filter fusion (`fusion/filter.rs`) fuses adjacent Filters into one Filter, which `CanonicalizeMfp` also does. This seems redundant, so we should consider eliminating one of these.

`CanonicalizeMfp` does more, so it should probably be the one to be retained. However, we should pay attention that  `canonicalize_predicates` is currently only called from filter fusion, so we will have to add a call in `CanonicalizeMfp`.

(Also, we might call `reduce` on the Map expressions in `CanonicalizeMfp`, while we are at it. But first check where else it's called.)

**Comment by ggevay** [comment]
I tried to do this, but there are several issues:
- `Demand` is causing an infinite fixpoint loop. Fortunately, this issue is resolved after rebasing to https://github.com/MaterializeInc/materialize/pull/17099 !
- There are several plan regressions where a non-trivial `Map` expression is no longer being absorbed (from below) into a `Reduce`. This is because `ReductionPushdown` can make a `Reduce` absorb a `Map`, but it's not smart enough to dig out a `Map` from below a `Filter`. This means that `CanonicalizeMfp` can make the situation worse from the point of view of `ReductionPushdown`, by switching the order of a `Map` and a `Filter`. Arguably, `CanonicalizeMfp` should never make things worse, i.e. when there is a regression after introducing a call to it, then some other code is to blame. So I think we should make `ReductionPushdown` slightly smarter by looking at a complete MFP rather than just a `Map`, and absorbing Map expressions from there (when the `Filter` is not referring to a `Map` expression).
- There is a regression for Filters that involve literal errors. This is because `CanonicalizeMfp` is calling `extract_non_errors_from_expr_mut`. It's unclear why it's avoiding extracting an MFP that includes errors, i.e., why it is not calling `extract_from_expression` (`mut`) instead. Maybe there is a good reason. This should be investigated before proceeding.

**Comment by frankmcsherry** [comment]
> Filter fusion (fusion/filter.rs) fuses adjacent Filters into one Filter, which CanonicalizeMfp also does. This seems redundant, so we should consider eliminating one of these.

If it's ok, I'd be happiest not causing this to happen yet. We have piles of things that are partially redundant with each other, and no firm foundation to stand on (for me, at least). For example, collection all of the "Fusion" tasks together seems more helpful in building that foundation than swapping some of them out for a transform that does several things to Map, Filter, and Project but only does fusion for a subset of the operators (e.g. not Union, Join, Negate, etc).

I'm hopeful that in the near future we end up with fewer confusing cases like you've listed, where weird and unpredictable interactions prevent us from moving forward with something that is meant to be a clear win!

**Comment by ggevay** [comment]
This was inspired by your [comment on Slack](https://materializeinc.slack.com/archives/C02PPB50ZHS/p1673529960750409?thread_ts=1673463352.216739&cid=C02PPB50ZHS), where you are suggesting to create/collect a normalization transform, and to call it very often. From there, my thinking went as follows:
1. `CanonicalizeMfp` totally sounds like it should be part of such a normalization transform.
2. This would mean that we add a call to `CanonicalizeMfp` in many places, probably also in `FuseAndCollapse`.
3. And then I noticed that it almost covers all the work that map fusion, filter fusion, and projection fusion do (which are a part of `FuseAndCollapse`), so then this looks like an opportunity to also eliminate some redundancy.

And then as a bonus, all the issues that were revealed when I actually tried to make this happen seem to be issues that should be fixed anyway, even if we were not replacing map, filter, projection fusions by `CanonicalizeMfp`, because these issues are other brittlenesses that can bite us at any time in any other optimizer work too.

> I'm hopeful that in the near future we end up with fewer confusing cases like you've listed, where weird and unpredictable interactions prevent us from moving forward with something that is meant to be a clear win!

Me too! :)

**Comment by ggevay** [comment]
Btw. I'm not planning to work on this in the near future, because the window functions work seems more urgent. So we can get back to this some time next month.

**Comment by frankmcsherry** [comment]
Sorry, I didn't mean to scare / randomize you. I should have been more clear that specifically the "remove the `fusion::Filter` transform" is the scary thing. For the moment, having both seems great, and when we get to remove e.g. `fusion::Filter` it would be great to replace it with a larger block that has a very specific behavior. I think .. there's a different world where we say "consistently normalizing MFP blocks is the main thing we need to do to clarify our transforms" in which case `CanonicalizeMFP` would be the best replacement, but the working hypothesis is that there is a different orientation of the foundational work (https://materializeinc.slack.com/archives/C02PPB50ZHS/p1673533834578579).

**Comment by ggevay** [comment]
> there is a different orientation of the foundational work

Ah, ok, so your different orientation would be to group transforms into blocks by what type of change they do (e.g., movement, fusion, canonicalization would be blocks). And this is at odds with what I tried to do here, because my grouping is more in the direction of grouping by operators: I tried to make `CanonicalizeMfp` be the thing that does all of movement, fusion, and canonicalization on just Map, Filters, and Projects (but not on other operators). Is my understanding correct?

**Comment by frankmcsherry** [comment]
I think so, admittedly not knowing exactly how much `CanonicalizeMFP` does. For example, I suspect it doesn't handle movement other than intra-{Map, Filter, Project} movement, e.g. not pushing them down through unions or lifting out of joins.

Tbh, exploring each of these approaches seems fine, and improving `CanonicalizeMFP` seems great as long as we don't at the same time delete stuff that would allow us to approach things a different way. Maybe said differently, deletion `fusion::Filter` would conflict with https://github.com/MaterializeInc/materialize/pull/17001. :D

**Comment by ggevay** [comment]
> I think so, admittedly not knowing exactly how much CanonicalizeMFP does. For example, I suspect it doesn't handle movement other than intra-{Map, Filter, Project} movement, e.g. not pushing them down through unions or lifting out of joins.

Yes, it doesn't consider any other operators, e.g. unions. It just looks at MFP blocks by themselves.

> deleting fusion::Filter would conflict with https://github.com/MaterializeInc/materialize/pull/17001. :D

Ahh, I didn't see this PR, sorry! I understand the problem now, thanks!

Ok, I'll get back to this after the window functions work.
